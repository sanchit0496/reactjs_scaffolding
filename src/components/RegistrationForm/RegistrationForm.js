// components/RegistrationForm.js
import React from 'react'
import { useFormik } from 'formik'
import * as Yup from 'yup'
import InputComponent from '../InputComponent/InputComponent'

const RegistrationForm = ({ onClose }) => {
  const formik = useFormik({
    initialValues: {
      firstName: '',
      lastName: '',
      age: '',
      email: '',
      password: '',
      confirmPassword: '',
    },
    validationSchema: Yup.object({
      firstName: Yup.string().required('Required'),
      lastName: Yup.string().required('Required'),
      age: Yup.number()
        .positive('Must be a positive number')
        .integer('Must be an integer')
        .required('Required'),
      email: Yup.string().email('Invalid email address').required('Required'),
      password: Yup.string()
        .min(8, 'Must be at least 8 characters')
        .required('Required'),
      confirmPassword: Yup.string()
        .oneOf([Yup.ref('password'), null], 'Passwords must match')
        .required('Required'),
    }),
    validateOnBlur: true, // Validate on blur
    validateOnChange: false, // Do not validate on change
    onSubmit: (values) => {
      // Handle form submission here
      console.log('Form submitted with values:', values)
      onClose() // Close the form after submission
    },
  })

  return (
    <form onSubmit={formik.handleSubmit}>
      <InputComponent
        label="First Name"
        id="firstName"
        name="firstName"
        type="text"
        value={formik.values.firstName}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.firstName && formik.errors.firstName}
      />

      <InputComponent
        label="Last Name"
        id="lastName"
        name="lastName"
        type="text"
        value={formik.values.lastName}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.lastName && formik.errors.lastName}
      />

      <InputComponent
        label="Age"
        id="age"
        name="age"
        type="text"
        value={formik.values.age}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.age && formik.errors.age}
      />

      <InputComponent
        label="Email"
        id="email"
        name="email"
        type="text"
        value={formik.values.email}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.email && formik.errors.email}
      />

      <InputComponent
        label="Password"
        id="password"
        name="password"
        type="password"
        value={formik.values.password}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.password && formik.errors.password}
      />

      <InputComponent
        label="Confirm Password"
        id="confirmPassword"
        name="confirmPassword"
        type="password"
        value={formik.values.confirmPassword}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.confirmPassword && formik.errors.confirmPassword}
      />

      <button type="submit">Submit</button>
    </form>
  )
}

export default RegistrationForm
