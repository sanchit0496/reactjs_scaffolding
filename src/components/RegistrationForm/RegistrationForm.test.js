import React from 'react'
import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import RegistrationForm from './RegistrationForm' // Adjust the path as necessary

describe('RegistrationForm', () => {
  // Snapshot Test
  it('renders correctly', () => {
    const { asFragment } = render(<RegistrationForm onClose={() => {}} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Functional Test - Input Changes and Validation
  it('updates input values and displays validation errors', async () => {
    render(<RegistrationForm onClose={() => {}} />)

    const firstNameInput = screen.getByLabelText('First Name')
    fireEvent.change(firstNameInput, { target: { value: 'John' } })
    expect(firstNameInput.value).toBe('John')

    const lastNameInput = screen.getByLabelText('Last Name')
    fireEvent.blur(lastNameInput)
    await waitFor(() => {
      expect(screen.getByText('Required')).toBeInTheDocument()
    })
  })

  // Functional Test - Form Submission
  it('submits the form', async () => {
    const mockOnClose = jest.fn()
    render(<RegistrationForm onClose={mockOnClose} />)

    // Fill out all required fields
    fireEvent.change(screen.getByLabelText('First Name'), {
      target: { value: 'John' },
    })
    fireEvent.change(screen.getByLabelText('Last Name'), {
      target: { value: 'Doe' },
    })
    fireEvent.change(screen.getByLabelText('Age'), { target: { value: '30' } })
    fireEvent.change(screen.getByLabelText('Email'), {
      target: { value: 'john@example.com' },
    })
    fireEvent.change(screen.getByLabelText('Password'), {
      target: { value: 'password123' },
    })
    fireEvent.change(screen.getByLabelText('Confirm Password'), {
      target: { value: 'password123' },
    })

    // Submit the form
    fireEvent.submit(screen.getByRole('button', { name: /Submit/i }))

    // Wait for onClose to be called
    await waitFor(() => {
      expect(mockOnClose).toHaveBeenCalled()
    })
  })

  // Additional tests can be added for each input field and validation scenario
})
