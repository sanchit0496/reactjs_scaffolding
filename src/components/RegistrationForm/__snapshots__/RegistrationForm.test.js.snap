// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm renders correctly 1`] = `
<DocumentFragment>
  <form>
    <div>
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="age"
      >
        Age
      </label>
      <input
        id="age"
        name="age"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div>
      <label
        for="confirmPassword"
      >
        Confirm Password
      </label>
      <input
        id="confirmPassword"
        name="confirmPassword"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
